pkg/virt-handler/vm.go

```diff
@@ -1851,51 +1851,15 @@ func (d *VirtualMachineController) processVmShutdown(vmi *v1.VirtualMachineInsta
 
 	// Only attempt to gracefully shutdown if the domain has the ACPI feature enabled
 	if isACPIEnabled(vmi, domain) {
-		expired, timeLeft := d.hasGracePeriodExpired(domain)
-		if !expired {
-			if domain.Status.Status != api.Shutdown {
-				err = client.ShutdownVirtualMachine(vmi)
-				if err != nil && !cmdclient.IsDisconnected(err) {
-					// Only report err if it wasn't the result of a disconnect.
-					return err
-				}
-
-				log.Log.Object(vmi).Infof("Signaled graceful shutdown for %s", vmi.GetObjectMeta().GetName())
-
-				// Make sure that we don't hot-loop in case we send the first domain notification
-				if timeLeft == -1 {
-					timeLeft = 5
-					if vmi.Spec.TerminationGracePeriodSeconds != nil && *vmi.Spec.TerminationGracePeriodSeconds < timeLeft {
-						timeLeft = *vmi.Spec.TerminationGracePeriodSeconds
-					}
-				}
-				// In case we have a long grace period, we want to resend the graceful shutdown every 5 seconds
-				// That's important since a booting OS can miss ACPI signals
-				if timeLeft > 5 {
-					timeLeft = 5
-				}
-
-				// pending graceful shutdown.
-				d.Queue.AddAfter(controller.VirtualMachineKey(vmi), time.Duration(timeLeft)*time.Second)
-				d.recorder.Event(vmi, k8sv1.EventTypeNormal, v1.ShuttingDown.String(), VMIGracefulShutdown)
-			} else {
-				log.Log.V(4).Object(vmi).Infof("%s is already shutting down.", vmi.GetObjectMeta().GetName())
-			}
-			return nil
+		if err, done := attemptShutdown(vmi, domain, d, client); !done {
```

Comment:
I don't like that the function has so many parameters, I think it would be better to split into more pieces to avoid passing all the params. 

For example, leaving the `d.hasGracePeriodExpired(domain)` and `if domain.Status.Status != api.Shutdown {` allows not passing domain at all, because it's not used any further